catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs

```diff
@@ -0,0 +1,328 @@
+/// 2. Determine entropy that was used to generate `mnemonic`.
+/// 3. Compute `pbkdf2_result` = PBKDF2-HMAC-SHA512(password = `passphrase`, salt =
+///    `entropy`, iterations = 4096, dkLen = 96).
+/// 4. given `pbkdf2_result` is S, modify S by assigning S[0] := S[0] & 0xf8 and S[31] :=
+///    (S[31] & 0x1f) | 0x40.
+/// 5. The result will be
+///     - kL where S[0:32] a 256-bit integer in little-endian byte order.
+///     - kR where S[32:64]
+///     - Result in (kL, kR) as the root extended private key and c := S[64:96] as the
+///       root chain code.
+fn mnemonic_to_xprv_helper(mnemonic: String, passphrase: Option<String>) -> anyhow::Result<XPrv> {
+    /// 4096 is the number of iterations for PBKDF2.
+    const ITER: u32 = 4096;
+
+    // Parse will detect language and check mnemonic valid length
+    // 12, 15, 18, 21, 24 are valid mnemonic length
+    let mnemonic =
+        Mnemonic::parse(mnemonic).map_err(|e| anyhow::anyhow!("Invalid mnemonic: {e}"))?;
+
+    let entropy = mnemonic.to_entropy();
+
+    let mut pbkdf2_result = [0; 96];
+    let _ = pbkdf2::<Hmac<Sha512>>(
+        passphrase.unwrap_or_default().as_bytes(),
+        &entropy,
+        ITER,
+        &mut pbkdf2_result,
+    );
+
+    Ok(XPrv::normalize_bytes_force3rd(pbkdf2_result))
+}
+
+/// Derive a new extended private key from the given extended private key.
+/// - V2 derivation scheme is used as it is mention in [SLIP-0023](https://github.com/satoshilabs/slips/blob/master/slip-0023.md).
+/// - More information about child key derivation can be found in [BIP32-Ed25519](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf).
+///  
+/// # Arguments
+///
+/// - `xprv_bytes`: An extended private key bytes of type `XPrvBytes`.
+/// - `path`: Derivation path. eg. m/0/2'/3 where ' represents hardened derivation.
+///
+/// # Returns
+///
+/// Returns a bytes of extended private key as a `Result`.
+///
+/// # Errors
+///
+/// Returns an error if the derivation path is invalid.
+// &str is not supported in flutter_rust_bridge
+#[allow(clippy::needless_pass_by_value)]
+pub async fn derive_xprv(xprv_bytes: XPrvBytes, path: String) -> anyhow::Result<XPrvBytes> {
+    let derive_xprv = spawn_blocking_with(
+        move || {
+            let xprv = XPrv::from_bytes_verified(xprv_bytes)?;
+            derive_xprv_helper(xprv, &path)
+        },
+        FLUTTER_RUST_BRIDGE_HANDLER.thread_pool(),
+    )
+    .await??;
+
+    Ok(derive_xprv.into())
+}
+
+/// Helper function for `derive_xprv`.
+fn derive_xprv_helper(xprv: XPrv, path: &str) -> anyhow::Result<XPrv> {
+    let Ok(derivation_path) = path.parse::<DerivationPath>() else {
+        return Err(anyhow::anyhow!("Invalid derivation path: {path}"));
+    };
+    let key = derivation_path.iter().fold(xprv, |xprv, child_num| {
+        if child_num.is_hardened() {
+            // i >= 2^31 is a hardened derivation
+            xprv.derive(DerivationScheme::V2, child_num.index() | 0x80_00_00_00)
```

Comment:
Can you explain what `0x80_00_00_00` means, please?

Reply:
`0x80_00_00_00` is equivalent to 2^31 which indicate harden derivation
Maybe this is more clear https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

Reply:
Can we please consider adding a comment?
Something like this:

```txt
// Apply hardened derivation by setting the highest bit (0x80000000).
// This ensures enhanced security by using a private derivation scheme,
// preventing parent key leakage even if child keys are compromised.
// Format: m/i' where i >= 0x80000000 (hardened index)
```
It could be worth adding the URL that you posted, too.